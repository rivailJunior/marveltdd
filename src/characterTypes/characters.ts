export type CharacterDataContainer = {
    offset: number
    limit: number
    total: number
    count: number
    results: [Character]
}

export type Character = {
    id: number
    name: string
    description: string
    modified: Date
    resourceURI: String
    urls?: string[] | null
    thumbnail?: Image
    comics?: ComicList[] | any
    stories?: StoryList[] | any
    events?: EventList[] | any
    series?: SeriesList[] | any
}

export type Image = {
    path: string
    extension: string
}

export type ComicList = {
    available: number
    returned: number
    collectionURI: string
    items: [ComicSummary]
}
export type ComicSummary = {
    resourceURI: string
    name: string
}

export type StoryList = {
    available: number
    returned: number
    collectionURI: string
    items: [StorySummary]
}

export type StorySummary = {
    resourceURI: string
    name: string
    type: string
}

export type EventList = {
    items?: EventSummary[] | null
    available: number
    returned: number
    collectionURI: string
}

export type EventSummary = {
    resourceURI: string
    name: string
}

export type SeriesList = {
    items?: SeriesSummary[] | null
    available: number
    returned: number
    collectionURI: string
}

export type SeriesSummary = {
    resourceURI: string
    name: string
}