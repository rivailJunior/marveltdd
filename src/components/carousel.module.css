.showCarousel {
    display: block;
}

.hideCarousel {
    display: none;
}