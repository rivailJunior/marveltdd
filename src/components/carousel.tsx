import { useState } from "react";
import { Carousel } from "react-bootstrap";
import styles from "./carousel.module.css";

type iCarouselProps = {
    data: [any];
    type: string;
};
export function CarouselInfor({ data, type }: iCarouselProps): JSX.Element {
    const [index, setIndex] = useState(0);

    const handleSelect = (selectedIndex: number, e: any) => {
        setIndex(selectedIndex);
    };

    const showErr = !data.length ? styles.showCarousel : styles.hideCarousel;
    const showCarousel = data.length
        ? styles.showCarousel
        : styles.hideCarousel;
    return (
        <>
            <div className={showErr}>Ops. Nenhuma {type} encontrado</div>
            <Carousel
                className={showCarousel}
                activeIndex={index}
                onSelect={handleSelect}
            >
                {data.map((item, index) => {
                    <Carousel.Item key={index}>
                        <img
                            className="d-block w-100"
                            src="holder.js/800x400?text=First slide&bg=373940"
                            alt="First slide"
                        />
                        <Carousel.Caption>
                            <h3>First slide label</h3>
                            <p>
                                Nulla vitae elit libero, a pharetra augue mollis
                                interdum.
                            </p>
                        </Carousel.Caption>
                    </Carousel.Item>;
                })}
            </Carousel>
        </>
    );
}
