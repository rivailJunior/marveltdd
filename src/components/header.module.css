.marvelHeader {
    /* padding: 10%; */
    top: 0px;
    left: 0px;
    width: 100%;
    height: 64px;
    background: #FFFFFF 0% 0% no-repeat padding-box;
    opacity: 1;
}

.labelName {
    text-align: left;
    font: normal normal bold 14px/20px PT Sans Caption;
    letter-spacing: 0px;
    color: #555555;
}

.informationProps {
    height: 100%;
    width: 100%;
    background-color: lightcoral;
}