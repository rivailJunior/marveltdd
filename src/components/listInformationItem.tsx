import styles from "./marvelList.module.css";

type iListProps = {
    list: [any] | any;
    tagName: string;
    information?: string;
};
export const ListInformationItem = ({
    list,
    tagName,
    information,
}: iListProps): JSX.Element => {
    return list?.length > 0 ? (
        <ul className={styles.list}>
            {list.slice(0, 3).map((item: any, index: number) => {
                return (
                    <li data-testid="informationLi" key={index}>
                        {item[tagName]}
                    </li>
                );
            })}
        </ul>
    ) : (
        <div className={styles.defaultLabel}>
            Ops. {information} n√£o encontrados!
        </div>
    );
};
