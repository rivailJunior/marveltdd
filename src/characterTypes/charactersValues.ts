
import { Character, CharacterDataContainer, ComicSummary, ComicList, StorySummary, StoryList, EventSummary, EventList, SeriesSummary, SeriesList, Image } from './characters';

export const ComicSummaryValue: ComicSummary = {
    resourceURI: "",
    name: ""
}

export const ImageValue: Image = {
    path: '',
    extension: ''
}

export const ComicListValue: ComicList = {
    available: 0,
    returned: 0,
    collectionURI: "",
    items: [ComicSummaryValue]
}

export const StorySummaryValue: StorySummary = {
    resourceURI: "",
    name: "",
    type: "",
}

export const StoryListValue: StoryList = {
    available: 0,
    returned: 0,
    collectionURI: "",
    items: [StorySummaryValue]
}

export const EventSummaryValue: EventSummary = {
    resourceURI: "",
    name: "",
}

export const EventListValue: EventList = {
    items: [EventSummaryValue],
    available: 0,
    returned: 0,
    collectionURI: "",
}

export const SeriesSummaryValue: SeriesSummary = {
    resourceURI: "",
    name: ""
}

export const SeriesListValue: SeriesList = {
    items: [SeriesSummaryValue],
    available: 0,
    returned: 0,
    collectionURI: ""

}

export const CharacterValue: Character = {
    id: 0,
    name: '',
    description: '',
    modified: new Date(),
    resourceURI: '',
    urls: [''],
    thumbnail: ImageValue,
    comics: [ComicListValue],
    stories: [StoryListValue],
    events: [EventListValue],
    series: [SeriesListValue]
}


export const CharacterDataContainerValue: CharacterDataContainer = {
    offset: 0,
    limit: 0,
    total: 0,
    count: 0,
    results: [CharacterValue]
}

