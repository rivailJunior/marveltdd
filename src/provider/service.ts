import { Character, CharacterDataContainer } from './../characterTypes/characters';
import axios, { AxiosResponse } from 'axios'
const apikey = process.env.REACT_APP_PUBKEY
// const hash = process.env.REACT_APP_HASH
// const ts = 1;

const api = axios.create({
    baseURL: 'https://gateway.marvel.com/v1/public',
});

const getService = async (url: string, options: object): Promise<AxiosResponse> => {
    return await api.get(url, { params: { apikey, ...options } })
}

const getCharacters = async (total: number): Promise<CharacterDataContainer> => {
    const result = await getService('/characters', { offset: 0, limit: total })
    return result.data?.data
}

const getParticipation = async (id: number, type: string): Promise<any> => {
    const result = await getService(`/characters/${id}/${type}`, {})
    return result.data?.data
}

const getCharacter = async (id: number): Promise<Character> => {
    const result = await getService(`/characters/${id}`, {})
    return result.data?.data?.results
}

export { getCharacters, getCharacter, getParticipation as getParticipations };