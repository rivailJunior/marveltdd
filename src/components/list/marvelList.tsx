import { Character } from "../../characterTypes/characters";
import { ListGroup, Col, Row, Image } from "react-bootstrap";
import styles from "./marvelList.module.css";
import { ListInformationItem } from "./listInformationItem";
import { useHistory } from "react-router-dom";

type iProps = {
    data: Character[];
    errInfor: boolean;
};

export const MarvelList = ({ data, errInfor }: iProps): JSX.Element => {
    let history = useHistory();

    const changePage = (item: Character | any) => {
        history.push("/description", item);
    };

    const showOrHideList = !errInfor ? styles.showList : styles.hideList;
    const showOrHideInformation = errInfor ? styles.showList : styles.hideList;
    return (
        <>
            <div className={"labelHeader " + showOrHideInformation}>
                Nenhum personagem encontrado
            </div>
            <ListGroup variant="flush" className={showOrHideList}>
                <ListGroup.Item style={{ backgroundColor: "transparent" }}>
                    <Row>
                        <Col md={2}></Col>
                        <Col md={2}>
                            <div className={styles.defaultLabel}>
                                Personagem
                            </div>
                        </Col>
                        <Col md={4} className="d-none d-lg-block">
                            <div className={styles.defaultLabel}>Séries</div>
                        </Col>
                        <Col md={4} className="d-none d-lg-block">
                            <div className={styles.defaultLabel}>Eventos</div>
                        </Col>
                    </Row>
                </ListGroup.Item>
            </ListGroup>
            <ListGroup className={styles.listGroup}>
                {data.map((character, index) => {
                    return (
                        <ListGroup.Item
                            key={index}
                            onClick={() => changePage(character)}
                            data-testid="marvelLi"
                        >
                            <Row>
                                <Col md={2} xs={6} lg={true}>
                                    <div className={styles.listImgDiv}>
                                        <Image
                                            className={styles.thumbImg}
                                            rounded
                                            src={`${ character?.thumbnail?.path }.${ character?.thumbnail?.extension }`}
                                            alt=""
                                        />
                                    </div>
                                </Col>
                                <Col md={2} xs={6} lg={true}>
                                    <div className={styles.defaultLabel}>
                                        {character.name}
                                    </div>
                                </Col>
                                <Col md={4} className="d-none d-lg-block">
                                    <ListInformationItem
                                        list={character.series?.items}
                                        tagName="name"
                                        information="Séries"
                                    />
                                </Col>
                                <Col md={4} className="d-none d-lg-block">
                                    <ListInformationItem
                                        list={character.events?.items}
                                        tagName="name"
                                        information="Eventos"
                                    />
                                </Col>
                            </Row>
                        </ListGroup.Item>
                    );
                })}
            </ListGroup>
        </>
    );
};
