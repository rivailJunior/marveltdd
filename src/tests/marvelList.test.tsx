import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { getCharacters } from "../provider/service";
import { MarvelList } from "../components/marvelList";
import { CharacterValue } from "../characterTypes/charactersValues";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";

afterEach(cleanup);

describe("`<MarvelList>`", () => {
    test("Render correctly", () => {
        const container = renderer.create(
            <MarvelList data={[CharacterValue]} />
        );
        container.toJSON();
        expect(container).toMatchSnapshot();
    });

    test("Render with no character list", () => {
        const { getByText } = render(<MarvelList data={[]} />);
        expect(getByText(/nenhum personagem encontrado/i)).toBeInTheDocument();
    });

    // test("Render with 10 characters in list", async () => {
    //     const result = await getCharacters(10);
    //     const { findAllByTestId } = render(
    //         <MarvelList data={result.results} />
    //     );
    //     const list = await findAllByTestId("marvelLi");
    //     expect(list).toHaveLength(10);
    // });

    // test("Call another page when click on list item", async () => {
    //     const result = await getCharacters(10);
    //     const { getAllByTestId, getByTestId } = render(
    //         <MemoryRouter initialEntries={["/my/initial/route"]}>
    //             <MarvelList data={result.results} />
    //         </MemoryRouter>
    //     );

    //     const btnShowDescription = getAllByTestId("marvelLi");
    //     fireEvent.click(btnShowDescription[0]);
    // });
});
