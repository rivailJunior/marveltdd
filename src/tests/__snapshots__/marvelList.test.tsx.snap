// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<MarvelList>\` Render correctly 1`] = `
Array [
  <div
    className="labelHeader hideList"
  >
    Nenhum personagem encontrado
  </div>,
  <div
    className="showList list-group list-group-flush"
    onKeyDown={[Function]}
  >
    <div
      className="list-group-item"
      data-rb-event-key={null}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <div
        className="row"
      >
        <div
          className="col-md-2"
        />
        <div
          className="col-md-2"
        >
          <div
            className="defaultLabel"
          >
            Personagem
          </div>
        </div>
        <div
          className="d-none d-lg-block col-md-4"
        >
          <div
            className="defaultLabel"
          >
            Séries
          </div>
        </div>
        <div
          className="d-none d-lg-block col-md-4"
        >
          <div
            className="defaultLabel"
          >
            Eventos
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="listGroup list-group"
    onKeyDown={[Function]}
  >
    <div
      className="list-group-item"
      data-rb-event-key={null}
      data-testid="marvelLi"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="col-lg col-md-2 col-6"
        >
          <div
            className="listImgDiv"
          >
            <img
              alt=""
              className="thumbImg rounded"
              src="."
            />
          </div>
        </div>
        <div
          className="col-lg col-md-2 col-6"
        >
          <div
            className="defaultLabel"
          >
            
          </div>
        </div>
        <div
          className="d-none d-lg-block col-md-4"
        >
          <div
            className="defaultLabel"
          >
            Ops. 
            Séries
             não encontrados!
          </div>
        </div>
        <div
          className="d-none d-lg-block col-md-4"
        >
          <div
            className="defaultLabel"
          >
            Ops. 
            Eventos
             não encontrados!
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
